testsuite/MDAnalysisTests/analysis/test_base.py

```diff
@@ -161,6 +161,31 @@ def test_AnalysisFromFunction(u, start, stop, step, nframes):
     for ana in (ana1, ana2, ana3):
         assert_equal(results, ana.results)
 
+        
+def mass_xyz(atomgroup, masses):
+        return atomgroup.positions * masses
+
+def test_AnalysisFromFunction_args_length(u):
+    protein = u.select_atoms('protein')
+    masses = protein.masses.reshape(-1, 1)
+    ans = base.AnalysisFromFunction(mass_xyz, protein, masses)
+    assert len(ans.args) == 2
+    assert int(sum(sum(sum(ans.run().results)))) == -317054
+
+def test_AnalysisFromFunction_args_content(u):
+    protein = u.select_atoms('protein')
+    masses = protein.masses.reshape(-1, 1)
+    ans = base.AnalysisFromFunction(mass_xyz, protein, masses)
+    assert (ans.args[0] == protein) and not (ans.args[1] - masses).any()
```

Comment:
`(ans.args[1] - masses).any()` tries to compare floating point numbers to 0. That's fragile. Avoid doing this.

Furthermore, what is going to happen if the code is still wrong? Are you able to calculate `ans.args[1] - masses` or will this throw a TypeError or something like that? In this case, this is not a clean failure. Please explain a bit more what the reasoning is behind this test.

Reply:
if the code is wrong, that is, `ans.args[0] is masses` and `ans.args[1] is protein`, will return `AssertionError`. Do we need to deal with this error, like `pytest.raises` ?

Reply:
`assert (ans.args[0] is protein) and (ans.args[1] is masses)` is better as you'll always get True/False answer and not an exception.

I'll call it a  "clean failure" when the `assert` raises `AssertionError` because of whatever you're asserting. Any other exception that gets raised will also make the test fail but in a way that we did not anticipate or explicitly test and it will show up differently when pytest runs. It's a sign of a bug, not a controlled test.